src/core/lib/security/credentials/credentials_metadata.c

```diff
@@ -24,65 +24,34 @@
 
 #include "src/core/lib/slice/slice_internal.h"
 
-static void store_ensure_capacity(grpc_credentials_md_store *store) {
-  if (store->num_entries == store->allocated) {
-    store->allocated = (store->allocated == 0) ? 1 : store->allocated * 2;
-    store->entries = gpr_realloc(
-        store->entries, store->allocated * sizeof(grpc_credentials_md));
+static void mdelem_list_ensure_capacity(grpc_credentials_mdelem_list *list,
+                                        size_t space_needed) {
+  size_t target_size = list->size + space_needed;
```

Comment:
Yeah, looks like the code wasn't quite right here.  The intent here is to keep the allocated space at the next power of two greater than the used space (i.e., every time we need more space, we double the allocated space).

The `size` field indicates the used space, not the allocated size.  The `space_needed` parameter is not the total space needed, but rather the additional space needed on top of the existing used space.  So the total space needed is the sum of the two.

So what we're doing here is to compute the next power of two greater than the new amount of space needed.  We then call `gpr_realloc()` to get that amount of space.

Note that it may be the case that the next power of two greater than the currently used space is the same as the next power of two greater than the new total space needed.  In this case, the call the `gpr_realloc()` will be a no-op.

I've tweaked the code to work the way I've described here and to make it a bit more readable.